### PR TITLE
Feature/ape 291 access control with scoped data

### PR DESCRIPTION
**Pull request by zashton**

### CODE REVIEW

File: app/Repositories/ProfileRepository.php [hunk 1]
```diff
@@ -9,44 +10,50 @@
...
+            if ($width > 240) {
+                $height = $height * 240 / $height;
+                $width = $width * 240 / $width;
             }
 
             $resized = $image
-                ->resize($width, $height, function ($constraint) { $constraint->aspectRatio(); } )
-                ->encode($file->getCLientOriginalExtension(),80);
-            $new_file_name = Str::slug(pathinfo(basename($file->getClientOriginalName()), PATHINFO_FILENAME)).'_'.time().'.'.$file->getCLientOriginalExtension();
+                ->resize($width, $height, function ($constraint) {
```
Comment (levity) [inline app/Repositories/ProfileRepository.php:49]: can you move this into its own PR please, so we can ship it sooner than the access control stuff? i also assigned the Linear task for this to you and marked it as in progress

File: database/migrations/2019_12_14_000001_create_personal_access_tokens_table.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('personal_access_tokens', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->morphs('tokenable');
```
[the remark this answers is not in the document]
Comment (zashton) [reply database/migrations/2019_12_14_000001_create_personal_access_tokens_table.php:18]: morphs is a `type`  column in the table, allows this record to be associated with any other table/model record via id. In this case issuing auth tokens to either `profile` , `circle`, `protocols` etc

File: app/Repositories/ProfileRepository.php (hunk 1 above)
Comment (zashton) [reply app/Repositories/ProfileRepository.php:49]: Haven't started on that task yet (could work on that first if its of more priority), this is probably because of the IDE doing its formatting as i've made some changes to this file

Comment (levity) [reply app/Repositories/ProfileRepository.php:49]: Ah, oops, I saw aspectRatio and jumped to conclusions

**Review by levity — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/CircleController.php [hunk 3]
```diff
@@ -6,41 +6,59 @@
...
         $circle = $this->repo->uploadCircleLogo($request);
-        if($circle)
+        if ($circle)
             return response()->json($circle);
 
-        return response()->json(['message'=> 'File Upload Failed' ,422]);
+        return response()->json(['message' => 'File Upload Failed', 422]);
     }
 
+    public function fullCircle(Request $request): JsonResponse
```
[the remark this answers is not in the document]
Comment (zashton) [reply app/Http/Controllers/CircleController.php]: renamed as well

File: app/Http/Controllers/ProfileController.php [hunk 4]
```diff
@@ -4,40 +4,93 @@
...
+        if (!$profile && count($profile->users))
+            abort('403', 'You do not have an active account in Coordinape');
+
+        if ($profile->admin_view) {
+            $circle_ids = ['*'];
+        } else {
+            $circle_ids = $profile->users->pluck('circle_id')->toArray();
+        }
+        $profile->tokens()->delete();
+        $token = $profile->createToken('circle-access-token', $circle_ids)->plainTextToken;
```
[the remark this answers is not in the document]
Comment (zashton) [reply app/Http/Controllers/ProfileController.php]: guess its because of the if statement is not checking whether the profile exists correctly
 ```
  if (!$profile && count($profile->users))
    abort('403', 'You do not have an active account in Coordinape');
```
replaced to this
https://github.com/coordinape/coordinape-backend/blob/5f3828cd02e0cc3265d60f64b1dc3a1f1987ee61/app/Http/Controllers/ProfileController.php#L78

File: app/Helper/Utils.php [hunk 5]
```diff
@@ -106,13 +107,43 @@ public static function queryCache($request,$callback,$minutes=1,$tags='default')
...
+        $profile = $request->user();
+        if ($profile) {
+            $query = $profile->users()->where('circle_id', $circle_id);
+            if ($is_admin) {
+                $query->isAdmin();
+            }
+            if ($address) {
+                $query->byAddress($address);
+            }
+            if ($admin_user = $query->first()) {
```
[the remark this answers is not in the document]
Comment (zashton) [reply app/Helper/Utils.php]: renamed this

File: app/Helper/Utils.php [hunk 6]
```diff
@@ -106,13 +107,43 @@ public static function queryCache($request,$callback,$minutes=1,$tags='default')
...
 
+    public static function getCircleUserByToken($request, $circle_id, $is_admin = false, $address = null)
+    {
+        $profile = $request->user();
+        if ($profile) {
+            $query = $profile->users()->where('circle_id', $circle_id);
+            if ($is_admin) {
+                $query->isAdmin();
+            }
+            if ($address) {
```
[the remark this answers is not in the document]
Comment (zashton) [reply app/Helper/Utils.php]: yeah you are right address totally is not necessary, since its all the same for the profile :sweat_smile:

File: app/Http/Controllers/ProfileController.php [hunk 7]
```diff
@@ -4,40 +4,93 @@
...
     }
 
-    public function getProfile(Request $request, $address) : JsonResponse {
-        return response()->json($this->repo->getProfile($request,$address));
+    public function getProfile(Request $request, $address): JsonResponse
+    {
+        $profile = $request->user();
+        $addressProfile = $this->repo->getProfile($address, ['users.circle.protocol', 'users.teammates', 'users.histories.epoch']);
+        if ($profile && !$profile->admin_view) {
+            if (!$addressProfile || count(array_intersect($profile->currentAccessToken()->abilities, $addressProfile->users()->pluck('circle_id')->toArray())) < 1) {
```
[the remark this answers is not in the document]
Comment (zashton) [reply app/Http/Controllers/ProfileController.php]: yeah, not by PHP but by laravel the token abilities is intentionally fixed with a cast that converts that column to an array. 
https://laravel.com/docs/8.x/eloquent-mutators#attribute-casting

File: app/Http/Controllers/ProfileController.php [hunk 8]
```diff
@@ -4,40 +4,93 @@
...
+    {
+        $profile = $this->repo->uploadProfileBackground($request);
+        if ($profile)
+            return response()->json($profile);
+
+        return response()->json(['message' => 'File Upload Failed', 422]);
+    }
+
+    // to be deprecated
+    public function uploadProfileAvatar(ProfileUploadRequest $request, $address): JsonResponse
```
[the remark this answers is not in the document]
Comment (zashton) [reply app/Http/Controllers/ProfileController.php]: removed this

File: app/Http/Controllers/NominationController.php [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 namespace App\Http\Controllers;
 
+use App\Helper\Utils;
```
Comment (levity) [inline app/Http/Controllers/NominationController.php:5]: this import is unused now

File: routes/api.php [hunk 10]
```diff
@@ -19,14 +20,88 @@
...
+// login
+Route::post('/v2/manifest', [ProfileController::class, 'manifest'])->middleware(['verify-sign']);
+
+Route::middleware(['verify-sign', 'hcaptcha-verify'])->group(function () {
+    Route::post('/v2/circles', [CircleController::class, 'createCircle']);
+});
+
+/************************* TOKEN GATED endpoints *********************************/
+
+Route::prefix('v2')->middleware(['auth:sanctum'])->group(function () {
```
Comment (levity) [inline routes/api.php:34]: ok, i think i understand now what i missed yesterday... having middleware(['auth:sanctum']) here means that all requests to routes in the group will fail if they don't have a valid token?

File: app/Http/Middleware/VerifyAdminToken.php [hunk 11]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace App\Http\Middleware;
+
+use App\Helper\Utils;
+use App\Models\User;
+use Closure;
+use Illuminate\Http\Request;
+
+class VerifyAdminToken
```
Comment (levity) [inline app/Http/Middleware/VerifyAdminToken.php:10]: since these two middlewares don't do anything with the token either, let's rename them too--something like `verify-circle-user`, `verify-circle-admin`

File: app/Repositories/ProfileRepository.php [hunk 12]
```diff
@@ -9,44 +10,50 @@
...
+        $profile->load(['users.circle.protocol', 'users.teammates', 'users.histories.epoch']);
         return $profile;
     }
 
-    public function uploadProfileAvatar($request, $address) {
-        $profile =  $this->model->byAddress($address)->first();
-        if($profile) {
+    public function uploadProfileAvatar($request, $address = null)
+    {
+        $profile = $address ? $this->model->byAddress($address)->first() : $request->user();
```
Comment (levity) [inline app/Repositories/ProfileRepository.php:36]: we don't allow uploading profile images for others, right? so it's ok to completely drop the $address arg

File: app/Helper/Utils.php [hunk 13]
```diff
@@ -106,13 +107,30 @@ public static function queryCache($request,$callback,$minutes=1,$tags='default')
...
+    {
+
+        return str_replace(array(':', '-', '/', '*', '_', '`'), ' ', $str);
 
-        return str_replace(array(':', '-', '/', '*','_','`'), ' ', $str);
+    }
+
+    public static function getCircleUserFromRequest($request, $circle_id, $is_admin = false)
+    {
+        $profile = $request->user();
```
Comment (levity) [inline app/Helper/Utils.php:123]: is it a problem that Sanctum is assigning a profile as `$request->user()`? does that conflict with the existing uses of `$request->user` at all?

**Review by levity — CHANGES_REQUESTED** [review]

File: app/Helper/Utils.php (hunk 13 above)
Comment (levity) [inline app/Helper/Utils.php:123]:
```suggestion
        $profile = $request->user(); // this is from Sanctum -- NOT the same as $request->user
```

File: routes/api.php (hunk 10 above)
Comment (zashton) [reply routes/api.php:34]: correct, it checks for the middleware at the outer most layer first, sanctum will throw an Unauthenticated error if theres no valid token inside the request header.

File: app/Http/Middleware/VerifyAdminToken.php (hunk 11 above)
Comment (zashton) [reply app/Http/Middleware/VerifyAdminToken.php]: updated

File: app/Repositories/ProfileRepository.php (hunk 12 above)
Comment (zashton) [reply app/Repositories/ProfileRepository.php]: removed this
